File: bakeoff/schema.py
"""v0.1 JSON Schema 校验 + 泄露检测(对齐 docs/CONTEXT.md §二、ADR-001 指标 #1/#3)。

两个校验入口:
  - validate_world(): world-gen 一次产出(world + character + rules + endings)。
  - validate_turn():  event-loop 单回合产出(narrative + 状态增量 + availableActions)。
泄露检测 detect_leak():扫描“玩家可见”文本,确认不含 isTrue / hiddenLogic 内容。
"""
from __future__ import annotations

from jsonschema import Draft202012Validator

# ── world-gen 产出(A 组)──────────────────────────────────────────────
WORLD_SCHEMA = {
    "type": "object",
    "required": ["schemaVersion", "mode", "archetypes", "world",
                 "character", "rules", "endings"],
    "additionalProperties": True,
    "properties": {
        "schemaVersion": {"const": "0.1"},
        "mode": {"enum": ["single", "hybrid"]},
        "archetypes": {"type": "array", "minItems": 1, "items": {"type": "string"}},
        "world": {
            "type": "object",
            "required": ["title", "background", "dangerLevel", "tone"],
            "properties": {
                "title": {"type": "string", "minLength": 1},
                "background": {"type": "string", "minLength": 1},
                "dangerLevel": {"enum": ["low", "medium", "high", "extreme"]},
                "tone": {"type": "string", "minLength": 1},
            },
        },
        "character": {
            "type": "object",
            "required": ["attributes"],
            "properties": {
                "attributes": {
                    "type": "object",
                    "required": ["hp", "san"],
                    "properties": {
                        "hp": {"type": "number", "minimum": 0, "maximum": 100},
                        "san": {"type": "number", "minimum": 0, "maximum": 100},
                    },
                },
                "traits": {"type": "array", "items": {"type": "string"}},
                "inventory": {"type": "array", "items": {"type": "string"}},
            },
        },
        "rules": {
            "type": "array",
            "minItems": 1,
            "items": {
                "type": "object",
                "required": ["id", "content", "isTrue", "hiddenLogic", "discovered"],
                "properties": {
                    "id": {"type": "integer"},
                    "content": {"type": "string", "minLength": 1},
                    "isTrue": {"type": "boolean"},
                    "hiddenLogic": {"type": "string"},
                    "discovered": {"type": "boolean"},
                },
            },
        },
        "endings": {
            "type": "array",
            "minItems": 1,
            "items": {
                "type": "object",
                "required": ["id", "title", "condition", "reached"],
                "properties": {
                    "id": {"type": "string"},
                    "title": {"type": "string", "minLength": 1},
                    "condition": {"type": "string", "minLength": 1},
                    "reached": {"type": "boolean"},
                },
            },
        },
    },
}

# ── event-loop 单回合产出(B 组)──────────────────────────────────────
# state 是真理之源、由引擎维护;模型每回合只回“叙事 + 增量 + 可选行动 + 结局判定”。
TURN_SCHEMA = {
    "type": "object",
    "required": ["narrative", "stateUpdate", "availableActions"],
    "additionalProperties": True,
    "properties": {
        "narrative": {"type": "string", "minLength": 1},
        "stateUpdate": {
            "type": "object",
            "required": ["hp", "san"],
            "properties": {
                "hp": {"type": "number", "minimum": 0, "maximum": 100},
                "san": {"type": "number", "minimum": 0, "maximum": 100},
                "timeline": {"type": "string"},
            },
        },
        # 本回合触发/新发现的规则 id(引擎据此置 discovered、做一致性核对)。
        "triggeredRuleIds": {"type": "array", "items": {"type": "integer"}},
        "discoveredRuleIds": {"type": "array", "items": {"type": "integer"}},
        "availableActions": {
            "type": "array",
            "minItems": 2,
            "maxItems": 4,
            "items": {
                "type": "object",
                "required": ["id", "text"],
                "properties": {
                    "id": {"type": "string"},
                    "text": {"type": "string", "minLength": 1},
                    "hint": {"type": "string"},
                },
            },
        },
        # 命中结局时返回 {reached: true, id: "..."},否则 null 或省略。
        "ending": {
            "type": ["object", "null"],
            "properties": {
                "reached": {"type": "boolean"},
                "id": {"type": "string"},
            },
        },
    },
}

_WORLD_V = Draft202012Validator(WORLD_SCHEMA)
_TURN_V = Draft202012Validator(TURN_SCHEMA)


def _errors(validator, obj) -> list[str]:
    return [
        f"{'/'.join(map(str, e.path)) or '<root>'}: {e.message}"
        for e in sorted(validator.iter_errors(obj), key=lambda e: list(e.path))
    ]


def validate_world(obj) -> list[str]:
    return _errors(_WORLD_V, obj)


def validate_turn(obj) -> list[str]:
    return _errors(_TURN_V, obj)


# ── 泄露检测(指标 #3:hiddenLogic / isTrue 绝不进玩家可见文本)──────────
LEAK_TOKENS = ("isTrue", "hiddenLogic", "isCorrect", "groundTruth")


def detect_leak(player_visible_text: str, world: dict | None = None) -> list[str]:
    """返回命中的泄露证据(空 = 干净)。

    1) 文本里出现引擎专用字段名(模型把内部结构吐给了玩家)。
    2) 文本里整段照抄了某条规则的 hiddenLogic 内容。
    """
    hits: list[str] = []
    text = player_visible_text or ""
    for tok in LEAK_TOKENS:
        if tok in text:
            hits.append(f"出现引擎字段名 {tok!r}")
    if world:
        for r in world.get("rules", []):
            hl = (r.get("hiddenLogic") or "").strip()
            if len(hl) >= 8 and hl in text:
                hits.append(f"照抄 rule#{r.get('id')} 的 hiddenLogic")
    return hits

File: web/src/types/schema.ts
// 统一 JSON Schema · v0.2 的 TypeScript 类型
// 真理之源是 docs/CONTEXT.md §二。本文件只定义类型,不写任何逻辑。
// 约定要点(v0.2,见 ADR-001 / bakeoff FINDINGS F-001):
//   - rules[].id 是【整数】,endings[].id 是【snake_case 字符串】,两者刻意不同。
//   - character.attributes 为【必填对象】,至少含该模式核心数值(规则怪谈: hp/san)。
//   - rules[].isTrue / hiddenLogic 是作者/引擎视角字段,绝不泄露给玩家。

export const SCHEMA_VERSION = '0.2' as const;

export type Mode = 'single' | 'hybrid';

export type Archetype =
  | 'rules_creepy' // 规则怪谈
  | 'life_sim' // 人生模拟
  | 'cultivation' // 修仙
  | 'cyberpunk' // 赛博朋克
  | 'apocalypse'; // 末日生存

export type DangerLevel = 'low' | 'medium' | 'high' | 'extreme';

export type GameStatus = 'ongoing' | 'ended';

export interface World {
  title: string;
  background: string;
  dangerLevel: DangerLevel;
  tone: string;
}

/** 数值字段因模式而异(规则怪谈: hp/san;修仙: 灵根/境界…)。默认 0–100。 */
export type CharacterAttributes = Record<string, number>;

export interface Character {
  /** 必填对象,至少含该模式的核心数值。 */
  attributes: CharacterAttributes;
  traits: string[];
  inventory: string[];
}

export interface Rule {
  /** 整数,便于引擎引用与状态回传。 */
  id: number;
  content: string;
  /** 作者/引擎视角,绝不泄露给玩家。 */
  isTrue: boolean;
  /** 作者/引擎视角,绝不泄露给玩家。 */
  hiddenLogic: string;
  discovered: boolean;
}

export interface LogEntry {
  turn: number;
  narrative: string;
  playerAction: string;
}

/** 每回合连同它一起回传模型的真理之源(注意与顶层对象区分)。 */
export interface GameState {
  turn: number;
  status: GameStatus;
  /** 当前世界线一句话摘要。 */
  timeline: string;
  /** 旧回合压缩后的摘要(成本控制)。 */
  logSummary: string;
  log: LogEntry[];
}

export interface AvailableAction {
  id: string;
  text: string;
  hint: string;
}

export interface Ending {
  /** snake_case 英文字符串,作为稳定语义标识。 */
  id: string;
  /** 短标题,必填。 */
  title: string;
  /** 整句结局描述,可选。 */
  description?: string;
  /** 可判定的中文条件。 */
  condition: string;
  reached: boolean;
}

/** 一局游戏的完整设定 + 状态(顶层对象)。state 字段才是每回合回传的真理之源。 */
export interface GameWorld {
  schemaVersion: string;
  mode: Mode;
  /** 1 个 = 单体,2–3 个 = 混合。 */
  archetypes: Archetype[];
  world: World;
  character: Character;
  rules: Rule[];
  state: GameState;
  availableActions: AvailableAction[];
  endings: Ending[];
}
